### DISCUSSION TITLE
how to use glsp action to send  this node data  to backend

### DISCUSSION THREAD

**Discussion by zrwsmd** [eclipse-glsp/glsp#1292] [Q&A]
with the basic of workflow example,i know how to create custom new node,and exexute save or other action like below 
![image](https://github.com/eclipse-glsp/glsp/assets/26995040/55540534-b82e-4c85-bd1f-cc95ac4add0c)

now my front end colleagues
drag a custom node,but just send me node data using sprotty LocalModelSource like below
![image](https://github.com/eclipse-glsp/glsp/assets/26995040/cf492363-7f37-4ab5-abbd-8dc500f0cfd2)

how to use glsp action to send  this node data  to backend,i mean which code can i trigger my custom node data to  execute related action operation(createNodeAction for example),in other word, how this node data pass to picture1 nodeOperation  param

**Comment by zrwsmd** [comment]
in summary,how could webview create a node and  send this node message to here
![image](https://github.com/eclipse-glsp/glsp/assets/26995040/fbfbf71a-0900-4ac1-b614-b82f58e57dc6)
at last i only can define  here,but how this msg pass to here 
![image](https://github.com/eclipse-glsp/glsp/assets/26995040/d711ac84-00c6-40a8-8058-31e0016e8986)

**Comment by planger** [comment]
Hi,

> drag a custom node,but just send me node data using sprotty LocalModelSource like below

Only sending data to the local model source will not work, as it would only update the diagram on the client and the server is unaware of this change. Thus, on the next model update from the server, the client state would be overwritten.

> how to use glsp action to send this node data to backend,i mean which code can i trigger my custom node data to execute related action operation(createNodeAction for example),in other word, how this node data pass to picture1 nodeOperation param

The correct way of informing the server about a change is to send an operation. See https://eclipse.dev/glsp/documentation/modeloperations/ for more details.

The server should process this operation then by applying it to the source model and regenerate the graph model, which is then sent as an update to the client after each operation. See https://eclipse.dev/glsp/documentation/sourcemodel/ for more details.

> in summary,how could webview create a node and send this node message to here

Every action (and operation, which is a type of action), for which a handler on the server is registered, will be forwarded to the server when sent by the client.

Best wishes,

Philip
